codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/RestJsonProtocolGenerator.java

```diff
@@ -44,55 +66,299 @@
     }
 
     @Override
-    protected void generateDocumentBodyShapeSerializers(GenerationContext context, Set<Shape> shapes) {
+    protected void generateOperationDocumentSerializer(
+            GenerationContext context,
+            OperationShape operation
+    ) {
+        Model model = context.getModel();
+        HttpBindingIndex bindingIndex = model.getKnowledge(HttpBindingIndex.class);
+        Set<MemberShape> documentBindings = bindingIndex.getRequestBindings(operation, HttpBinding.Location.DOCUMENT)
+                .stream()
+                .map(HttpBinding::getMember)
+                .collect(Collectors.toSet());
+
+        if (documentBindings.size() == 0) {
+            return;
+        }
+
+        Shape inputShape = model.expectShape(operation.getInput()
+                .orElseThrow(() -> new CodegenException("Input shape missing for operation " + operation.getId())));
+        GoWriter writer = context.getWriter();
+
+        writeJsonShapeSerializerFunction(writer, model, context.getSymbolProvider(), inputShape,
+                documentBindings::contains);
+        writer.write("");
     }
 
-    @Override
-    protected void generateDocumentBodyShapeDeserializers(GenerationContext context, Set<Shape> shapes) {
+    private void writeJsonShapeSerializerFunction(
+            GoWriter writer,
+            Model model,
+            SymbolProvider symbolProvider,
+            Shape inputShape,
+            Predicate<MemberShape> filterMemberShapes
+    ) {
+        Symbol jsonEncoder = SymbolUtils.createPointableSymbolBuilder("Value", GoDependency.AWS_JSON_PROTOCOL)
+                .build();
+        Symbol inputSymbol = symbolProvider.toSymbol(inputShape);
+        String functionName = ProtocolGenerator.getDocumentSerializerFunctionName(inputShape, getProtocolName());
+
+        writer.addUseImports(GoDependency.FMT);
```

Comment:
The code in this PR could use a decent chunk of comments, especially for big sections like this, just for general "here's what we're doing (and possibly why)" notes.